SimTracker/SiPixelDigitizer/plugins/SiPixelChargeReweightingAlgorithm.cc

```diff
@@ -0,0 +1,598 @@
+#include "CondFormats/SiPixelObjects/interface/SiPixelLorentzAngle.h"
+#include "CondFormats/SiPixelObjects/interface/SiPixelQuality.h"
+#include "CondFormats/SiPixelObjects/interface/PixelROC.h"
+#include "CondFormats/SiPixelObjects/interface/LocalPixel.h"
+#include "CondFormats/SiPixelObjects/interface/CablingPathToDetUnit.h"
+#include "CondFormats/SiPixelObjects/interface/SiPixelDynamicInefficiency.h"
+#include "CondFormats/SiPixelObjects/interface/SiPixelFEDChannelContainer.h"
+#include "CondFormats/SiPixelObjects/interface/SiPixelQualityProbabilities.h"
+#include "CondFormats/SiPixelObjects/interface/SiPixel2DTemplateDBObject.h"
+
+#include "CondFormats/SiPixelObjects/interface/SiPixelFrameReverter.h"
+#include "CondFormats/SiPixelObjects/interface/PixelFEDCabling.h"
+#include "CondFormats/SiPixelObjects/interface/PixelFEDLink.h"
+#include "DataFormats/FEDRawData/interface/FEDNumbering.h"
+#include "SimDataFormats/PileupSummaryInfo/interface/PileupMixingContent.h"
+#include "SimDataFormats/Track/interface/SimTrack.h"
+
+// Geometry
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+
+#include "CondFormats/SiPixelObjects/interface/PixelROC.h"
+
+using namespace edm;
+using namespace sipixelobjects;
+
+void SiPixelChargeReweightingAlgorithm::init(const edm::EventSetup& es) {
+  // Read template files for charge reweighting
+  if (UseReweighting) {
+    edm::ESHandle<SiPixel2DTemplateDBObject> SiPixel2DTemp_den;
+    es.get<SiPixel2DTemplateDBObjectRcd>().get("denominator", SiPixel2DTemp_den);
+    dbobject_den = SiPixel2DTemp_den.product();
+
+    edm::ESHandle<SiPixel2DTemplateDBObject> SiPixel2DTemp_num;
+    es.get<SiPixel2DTemplateDBObjectRcd>().get("numerator", SiPixel2DTemp_num);
+    dbobject_num = SiPixel2DTemp_num.product();
+
+    int numOfTemplates = dbobject_den->numOfTempl() + dbobject_num->numOfTempl();
+    templateStores_.reserve(numOfTemplates);
+    SiPixelTemplate2D::pushfile(*dbobject_den, templateStores_);
+    SiPixelTemplate2D::pushfile(*dbobject_num, templateStores_);
+
+    track.resize(6);
+  }
+}
+
+//=========================================================================
+
+SiPixelChargeReweightingAlgorithm::SiPixelChargeReweightingAlgorithm(const edm::ParameterSet& conf)
+    :
+
+      templ2D(templateStores_),
+      xdouble(TXSIZE),
+      ydouble(TYSIZE),
+      IDnum(conf.exists("TemplateIDnumerator") ? conf.getParameter<int>("TemplateIDnumerator") : 0),
+      IDden(conf.exists("TemplateIDdenominator") ? conf.getParameter<int>("TemplateIDdenominator") : 0),
+
+      UseReweighting(conf.getParameter<bool>("UseReweighting")),
+      PrintClusters(conf.getParameter<bool>("PrintClusters")),
+      PrintTemplates(conf.getParameter<bool>("PrintTemplates")) {
+  LogInfo("PixelDigitizer ") << "SiPixelChargeReweightingAlgorithm constructed"
```

Comment:
@carolinecollard , better to use edm::LogVerbatim

Reply:
ok. I am changing it. I will update on git in a few minutes.